File: tests/generate/apdu.ml
(* Copyright 2023 Nomadic Labs <contact@nomadic-labs.com>
   Copyright 2023 Functori <contact@functori.com>

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License. *)

module Class = struct
  type t = Default

  let to_uint8 = function Default -> 0x80
end

module Instruction = struct
  type t =
    | Version
    | Authorize_baking
    | Get_public_key
    | Prompt_public_key
    | Sign
    | Sign_unsafe
    | Reset
    | Query_auth_key
    | Query_main_hwm
    | Git
    | Setup
    | Query_all_hwm
    | Deauthorize
    | Query_auth_key_with_curve
    | Hmac
    | Sign_with_hash

  let to_uint8 = function
    | Version -> 0x00
    | Authorize_baking -> 0x01
    | Get_public_key -> 0x02
    | Prompt_public_key -> 0x03
    | Sign -> 0x04
    | Sign_unsafe -> 0x05
    | Reset -> 0x06
    | Query_auth_key -> 0x07
    | Query_main_hwm -> 0x08
    | Git -> 0x09
    | Setup -> 0x0A
    | Query_all_hwm -> 0x0B
    | Deauthorize -> 0x0C
    | Query_auth_key_with_curve -> 0x0D
    | Hmac -> 0x0E
    | Sign_with_hash -> 0x0F
end

module Index = struct
  type t = int

  let to_uint8 ?(is_last = false) idx =
    let is_last = if is_last then 0x80 else 0x00 in
    is_last lor idx
end

module Curve = struct
  type t = ED25519 | SECP256K1 | SECP256R1 | BIP32_ED25519

  let of_sk : Tezos_crypto.Signature.secret_key -> t = function
    | Ed25519 _ -> ED25519
    | Secp256k1 _ -> SECP256K1
    | P256 _ -> SECP256R1
    | Bls _ -> failwith "The Bls curve is not supported"

  let to_uint8 = function
    | ED25519 -> 0x00
    | SECP256K1 -> 0x01
    | SECP256R1 -> 0x02
    | BIP32_ED25519 -> 0x03

  let deterministic_sig = function
    | ED25519 -> true
    | SECP256K1 | SECP256R1 | BIP32_ED25519 -> false
end

let success = Bytes.of_string "\x90\x00"
let max_size = 235

let make_packet ?is_last ?(idx = 0) ~cla ~ins ~curve content =
  let content_length = Bytes.length content in
  assert (content_length <= max_size);
  let header = Bytes.create 5 in
  Bytes.set_uint8 header 0 @@ Class.to_uint8 cla;
  Bytes.set_uint8 header 1 @@ Instruction.to_uint8 ins;
  Bytes.set_uint8 header 2 @@ Index.to_uint8 ?is_last idx;
  Bytes.set_uint8 header 3 @@ Curve.to_uint8 curve;
  Bytes.set_uint8 header 4 content_length;
  Bytes.cat header content

let make_packets ?(idx = 0) ~cla ~ins ~curve content =
  let len = Bytes.length content in
  let rec aux ofs idx =
    let remaining_size = len - ofs in
    if remaining_size <= max_size then
      let packet =
        make_packet ~is_last:true ~idx ~cla ~ins ~curve
        @@ Bytes.sub content ofs remaining_size
      in
      [ packet ]
    else
      let packet =
        make_packet ~idx ~cla ~ins ~curve @@ Bytes.sub content ofs max_size
      in
      packet :: aux (ofs + max_size) (idx + 1)
  in
  aux 0 idx

module Path = struct
  type t = int list

  let to_bytes path =
    (* hardened is defined by the "'" *)
    let hardened idx = idx lor 0x80_00_00_00 in
    let length = List.length path in
    let bytes = Bytes.create ((4 * length) + 1) in
    Bytes.set_uint8 bytes 0 length;
    List.iteri
      (fun i idx ->
        Bytes.set_int32_be bytes ((4 * i) + 1) (Int32.of_int (hardened idx)))
      path;
    bytes
end

module Signer = struct
  type t = {
    path : Path.t;
    pkh : Tezos_crypto.Signature.public_key_hash;
    pk : Tezos_crypto.Signature.public_key;
    sk : Tezos_crypto.Signature.secret_key;
  }

  let make ~path ~sk =
    let open Tezos_crypto.Signature in
    let sk = Secret_key.of_b58check_exn sk in
    let pk = Secret_key.to_public_key sk in
    let pkh = Public_key.hash pk in
    { path; pkh; pk; sk }
end

File: tests/integration/nano/utils/apdu.py
# Copyright 2023 Functori <contact@functori.com>

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at

# http://www.apache.org/licenses/LICENSE-2.0

# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from enum import IntEnum
from typing import Union

from ragger.backend.interface import BackendInterface, RAPDU
from ragger.bip import pack_derivation_path
from ragger.error import ExceptionRAPDU

class CLA(IntEnum):
    DEFAULT = 0x80

class INS(IntEnum):
    VERSION                   = 0x00
    AUTHORIZE_BAKING          = 0x01
    GET_PUBLIC_KEY            = 0x02
    PROMPT_PUBLIC_KEY         = 0x03
    SIGN                      = 0x04
    SIGN_UNSAFE               = 0x05
    RESET                     = 0x06
    QUERY_AUTH_KEY            = 0x07
    QUERY_MAIN_HWM            = 0x08
    GIT                       = 0x09
    SETUP                     = 0x0a
    QUERY_ALL_HWM             = 0x0b
    DEAUTHORIZE               = 0x0c
    QUERY_AUTH_KEY_WITH_CURVE = 0x0d
    HMAC                      = 0x0e
    SIGN_WITH_HASH            = 0x0f

class INDEX(IntEnum):
    FIRST      = 0x00
    OTHER      = 0x01
    LAST       = 0x80
    OTHER_LAST = 0x81

class SIGNATURE_TYPE(IntEnum):
    ED25519       = 0x00
    SECP256K1     = 0x01
    SECP256R1     = 0x02
    BIP32_ED25519 = 0x03

class StatusCode(IntEnum):
    SECURITY                  = 0x6982
    HID_REQUIRED              = 0x6983
    REJECT                    = 0x6985
    WRONG_VALUES              = 0x6A80
    REFERENCED_DATA_NOT_FOUND = 0x6A88
    WRONG_PARAM               = 0x6B00
    WRONG_LENGTH              = 0x6C00
    INVALID_INS               = 0x6D00
    CLASS                     = 0x6E00
    OK                        = 0x9000
    UNEXPECTED_STATE          = 0x9001
    UNEXPECTED_SIGN_STATE     = 0x9002
    UNKNOWN_CX_ERR            = 0x9003
    UNKNOWN                   = 0x90FF
    WRONG_LENGTH_FOR_INS      = 0x917E
    MEMORY_ERROR              = 0x9200
    PARSE_ERROR               = 0x9405

class APP_KIND(IntEnum):
    WALLET = 0x00
    BAKING = 0x01

class Account:
    def __init__(self,
                 path: Union[str, bytes],
                 sig_type: SIGNATURE_TYPE,
                 public_key: str):
        self.path: bytes = \
            pack_derivation_path(path) if isinstance(path, str) \
            else path
        self.sig_type: SIGNATURE_TYPE = sig_type
        self.public_key: str = public_key

MAX_APDU_SIZE: int = 235

class TezosBackend(BackendInterface):

    def _exchange(self,
                  ins: INS,
                  index: INDEX = INDEX.FIRST,
                  sig_type: Union[SIGNATURE_TYPE, None] = None,
                  payload: bytes = b'') -> bytes:

        assert len(payload) <= MAX_APDU_SIZE, f"Apdu too large"

        # Set to a non-existent value to ensure that p2 is unused
        p2: int = sig_type if sig_type is not None else 0xff

        rapdu: RAPDU = self.exchange(CLA.DEFAULT,
                                     ins,
                                     p1=index,
                                     p2=p2,
                                     data=payload)

        if rapdu.status != StatusCode.OK:
            raise ExceptionRAPDU(rapdu.status, rapdu.data)

        return rapdu.data

    def git(self) -> bytes:
        return self._exchange(INS.GIT)

    def version(self) -> bytes:
        return self._exchange(INS.VERSION)

    def get_public_key(self,
                       account: Account,
                       with_prompt: bool = False) -> bytes:

        ins = INS.PROMPT_PUBLIC_KEY if with_prompt else INS.GET_PUBLIC_KEY

        return self._exchange(ins,
                              sig_type=account.sig_type,
                              payload=account.path)

    def _ask_sign(self, ins: INS, account: Account) -> None:
        data: bytes = self._exchange(ins, sig_type=account.sig_type, payload=account.path)
        assert not data

    def _continue_sign(self, ins: INS, message: Union[str, bytes], last: bool) -> bytes:
        if isinstance(message, str): message = bytes.fromhex(message)
        index: INDEX = INDEX.OTHER
        if last: index = INDEX(index | INDEX.LAST)
        return self._exchange(ins, index, payload=message)

    def sign(self,
             account: Account,
             message: Union[str, bytes],
             with_hash: bool = False,
             apdu_size: int = MAX_APDU_SIZE) -> bytes:
        if isinstance(message, str): message = bytes.fromhex(message)
        assert message, "Do not sign empty message"

        ins = INS.SIGN_WITH_HASH if with_hash else INS.SIGN

        self._ask_sign(ins, account)

        while(message):
            payload = message[:apdu_size]
            message = message[apdu_size:]
            last = not message
            data = self._continue_sign(ins, payload, last)
            if last:
                return data
            else:
                assert not data

        assert False, "We should have already returned"
